### DISCUSSION TITLE
overlaying streamlines on image:  ValueError: "ERROR:root:Algorithm

### DISCUSSION THREAD

**Discussion by gabrielDiaz-performlab** [pyvista/pyvista#3821] [Q&A]
Hi folks,

New to vista.  I would like to use streamlines in a way that was not originally intended, but that should be very possible.  In another context, I have extracted optical flow estimates from a movie (e.g., for every frame of the movie, I have a per-pixel velocity vector).  For each frame in the movie, I would like the frame's optical flow field to act as my vectors when making a streamlines figure overlay on the underlying image.  Of course, I'm starting with a single image / flow field pairing to debug. 

I would not be the first to do this.  Here is an existing example of my intended output:
https://www.youtube.com/watch?v=36DwddZ8EKM&ab_channel=JonMatthis

I am able to create a mesh from a greyscale image that was stored in a numpy array.  I have added velocity vectors.  However, when I try to run streamlines_evenly_spaced_2D(), I get the following error:

**"ValueError: "ERROR:root:Algorithm vtkEvenlySpacedStreamlines2D(0000022ECAFD0430) returned failure for request: vtkInformation (0000022EAEA9E7B0)"**

A related question:  Is there a better data representation of the image on which to overlay my streamlines?  I'm not quite sure what the strategy will be for saving out the rasterized image when I'm done, but I have to imagine there's a solution.   Should I be adding my image as a texture, rather than point data?  

My goal is to export the streamlines (perhaps overlaid on the original image data) as an image or image array of the same dimensions as my original image (e.g. 480 x 640 x rgb).    

Open to feedback.

Here is a link to the streamlines.pickle file references in the code.  It was pickled using Py 3.8, and you may have trouble if using a different version of pickle to unpickle.
https://drive.google.com/file/d/1YUHB3wOV-JaJ5ywhA8eBdW7b-RhsyTXn/view?usp=share_link

Here is my code.

```
import pickle
import numpy as np
import matplotlib.pyplot as plt
import pyvista as pv
import os
import sys
# The next 2 lines should be commented out if anyone copies this code.  This is b/c I have Cuda support for opencv, and it's missing a 
# dependency or two.
os.add_dll_directory("D://opencvgpu//opencv_build_310//bin")
os.add_dll_directory("C://Program Files//NVIDIA GPU Computing Toolkit//CUDA//v11.8//bin")

sys.path.append('core')

import cv2

fid = open("streamlines.pickle",'rb')
data = pickle.load(fid)

flow  = data['flow'] # a (480, 640, 2) vector field (optical flow)
frame = data['frame'] # a (480, 640, 3) image in bgr format
frame = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB) # a (480, 640, 3) image in rgb format

w = int(np.shape(frame)[1])
h = int(np.shape(frame)[0])

x = np.arange(0, np.shape(frame)[0], 1)
y = np.arange(0, np.shape(frame)[1], 1)
xx, yy, zz = np.meshgrid(y, x, 0)

grid = pv.UniformGrid()
grid.dimensions = (w,h,1)
grid.origin = (int(w/2),int(h/2),0)
grid.spacing = (1,1,1)
grid.show_scalar_bar = False

grid.texture_map_to_plane(inplace=True)
tex = pv.numpy_to_texture(cv2.rotate(frame, cv2.ROTATE_180))


flow3D = np.zeros([np.shape(flow)[0]*np.shape(flow)[1],3],dtype=np.float32)
flow3D[:,0] = flow[...,0].flatten()
flow3D[:,1] = flow[...,1].flatten()
grid['vectors'] = flow3D
grid.set_active_vectors("vectors") # > ValueError: Data field (flow) with type (FieldAssociation.NONE) not usable

line_streamlines = grid.streamlines_evenly_spaced_2D(
    start_position=(0, 0, 0.0),
    separating_distance=3,
    separating_distance_ratio=0.2,
    compute_vorticity=False,  # vorticity already exists in dataset
)

grid.plot(texture=tex)
```

Perhaps this will also help : )

![image](https://user-images.githubusercontent.com/8962011/212168620-e3d64fe1-1096-4b18-94c2-5c06bf07dffe.png)

**Comment by MatthewFlamm** [comment]
Edit:  This response looks like it was more relevant for a previous edit of your question.

Since your vector data exists also on the cells, you can load that data into the cell data too:

```py
flow3D = np.zeros([np.shape(flow)[0]*np.shape(flow)[1],3],dtype=np.float32)
flow3D[:,0] = flow[...,0].flatten()
flow3D[:,1] = flow[...,0].flatten()
grid.cell_data("flow") = flow3D  # Note difference here
```

I _think_ you will need to use point data for the streamlines, although it seems this isn't well documented.  I recommend starting with larger values for `separating_distance` and so on until you have a stable workflow.

```py
grid_point = grid.cell_data_to_point_data()
line_streamlines = grid_point.streamlines_evenly_spaced_2D(
    start_position=(0, 0, 0.0),
    separating_distance=3,
    separating_distance_ratio=0.2,
```

**Reply by gabrielDiaz-performlab** [reply]
Thank you for the very rapid response.  In fact, while you were responding I was able to add active vector data.  However, I now get another error.  See my (now modified) comment above.

To be thorough, I tried your suggestion, but the vectors for flow3D do not seem to be the correct size.

grid.cell_data["flow"] =  flow3D  
# ValueError: data length of (307200) != required length (306081)

My vectors = the number of cells.  They apparently want data equal to the number of points to add vectors.  Does this provide more insight into the underlying issue?

**Reply by MatthewFlamm** [reply]
No this means I was mistaken and that the length of your vectors == the number of points.  Check your numbers again.  I will respond in another thread to something else you can try.

**Comment by MatthewFlamm** [comment]
`"ValueError: "ERROR:root:Algorithm vtkEvenlySpacedStreamlines2D(0000022ECAFD0430) returned failure for request: vtkInformation (0000022EAEA9E7B0)"`

I don't know if this was ever tested with UniformGrid (Edit: I tested this with a simple case and it works fine, so I don't think this is it).  I'm a little hesitant to download and use pickled data, so I can't try your data.  You _might_ be running into a situation where you data is way too noisy for streamlines_evenly_spaced_2D.  This filter can be unstable and run out of memory quickly (although I'm not sure that is your problem yet).  First try streamlines with a single starting point and build up complexity from there.  https://docs.pyvista.org/examples/01-filter/streamlines_2D.html

**Reply by gabrielDiaz-performlab** [reply]
No luck.  A single streamline does not error, but neither does it return a valid mesh.
![image](https://user-images.githubusercontent.com/8962011/212173959-4b081c03-7edf-4587-a8ed-ce5e3547ecc7.png)

I've tried a few settings.   Any other ideas?  I'm a bit lost at this point - my understanding is that my data is prepared correctly but, as you suggested, the docs aren't full enough to help me out here.

Love what I've seen about pyvista, though.  :)

**Reply by MatthewFlamm** [reply]
It is my experience that this happens in a couple of scenarios: 

 - Your seed point could be outside the domain.  From your original screenshot, your starting position is not within the x and y bounds from this latest screenshot.
 - Your vector data is so noisy that there is no successful integration steps at this point and no streamline is generated.

**Reply by gabrielDiaz-performlab** [reply]
You were right - I was not using bounds correctly.  I made good progress, but right now it appears that pyvista's streamlines_evenly_spaced_2D is very, very unstable, and more often than not crashes my program on very small videos (e.g. 100 frames or less).  No errors are thrown, and try/except isn't able to catch the issue.  

Thanks for your continued support.  If you have any other ideas, I'm willing to try them.  The speed that this operates is unparalleled.  

![image](https://user-images.githubusercontent.com/8962011/212200910-74e859c9-9d2b-4a4c-9d0d-677547c8eb4c.png)


Here is the code I was working with.  I have played with the parameters of streamlines_evenly_spaced_2D .  It's success is very sensitive to those settings.

```

        # dbfile = open(os.path.join(self.video_out_path,"streamlines.pickle"), 'wb')
        # pickle.dump({"frame": frame, "flow": flow},dbfile)
        # dbfile.close()

        w = int(np.shape(frame)[1])
        h = int(np.shape(frame)[0])

        x = np.arange(0, np.shape(frame)[0], 20)
        y = np.arange(0, np.shape(frame)[1], 20)
        xx, yy, zz = np.meshgrid(y, x, 0)

        grid = pv.UniformGrid()
        grid.dimensions = (w, h, 1)
        grid.origin = (0, 0, 0)
        grid.spacing = (1, 1, 1)
        grid.show_scalar_bar = False

        grid.texture_map_to_plane(inplace=True)
        # tex = pv.numpy_to_texture(cv2.rotate(frame, cv2.ROTATE_180))
        tex = pv.numpy_to_texture(frame)

        flow3D = np.zeros([np.shape(flow)[0] * np.shape(flow)[1], 3], dtype=np.float32)
        flow3D[:, 0] = -flow[..., 0].flatten()
        flow3D[:, 1] = -flow[..., 1].flatten()

        grid['vectors'] = flow3D
        grid.set_active_vectors("vectors")

        p = pv.Plotter(off_screen=True, window_size=[640, 480])
        p.add_mesh(grid, texture=tex)

        start_pts = np.array(
            [[x, y, 0] for x in np.arange(0, np.shape(frame)[0], 20) for y in np.arange(0, np.shape(frame)[1], 20)])

        try:

            line_streamlines = grid.streamlines_evenly_spaced_2D(
                start_position=(0, 0, 0),
                step_length=10,
                separating_distance=20,
                separating_distance_ratio=0.7,
                compute_vorticity=False,  # vorticity already exists in dataset
                vectors="vectors",
                #     progress_bar= True,
            )
        except:
            print('B')
            return frame

        try:
            p.add_mesh(line_streamlines.tube(radius=2), color='w')
        except:
            print('C')
            return frame
        
        p.camera_position = 'xy'
        p.view_xy()
        p.camera.up = (0.0, 1.0, 0.0)
        p.camera.zoom(1.7)
        p.store_image=True
        p.render()
        p.show()
        
        im = p.image
        p.close()

    
        return cv2.cvtColor(im, cv2.COLOR_RGB2BGR)



```

**Reply by MatthewFlamm** [reply]
Looks great!  In my experience `streamlines_evenly_spaced_2D` is too unstable for repeated use over and over.  I can find parameters that work for a specific dataset with data, but may have to adjust them on a slightly different dataset.

**Reply by MatthewFlamm** [reply]
> No errors are thrown, and try/except isn't able to catch the issue.

The heavy lifting is done in vtk which is c++ so it won't necessarily be caught in python like you'd expect.
